### PR TITLE
fix(cli): fixes for server bundling

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Enable package exports for server code.
- Add missing babel plugin for server.
- Copy shims in case of working in a monorepo.
- Support `src/app` directory with API routes.

# How

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/start/server/metro/externals.ts [hunk 1]
```diff
@@ -31,6 +34,14 @@ export function getNodeExternalModuleId(fromModule: string, moduleId: string) {
   );
 }
 
+export async function setupShimFiles(projectRoot: string) {
+  await fs.promises.mkdir(path.join(projectRoot, METRO_SHIMS_FOLDER), { recursive: true });
+  // Copy the shims to the project folder in case we're running in a monorepo.
+  const shimsFolder = path.join(require.resolve('@expo/cli/package.json'), '..', 'static/shims');
```
Comment (ide) [inline packages/@expo/cli/src/start/server/metro/externals.ts:40]: prefer path.dirname over using .. to get the directory from a file path

File: packages/babel-preset-expo/src/index.ts [hunk 2]
```diff
@@ -112,6 +112,10 @@ function babelPresetExpo(api: ConfigAPI, options: BabelPresetExpoOptions = {}):
     extraPlugins.push(expoRouterBabelPlugin);
   }
 
+  if (api.caller((caller) => caller?.isServer)) {
+    extraPlugins.push(require('@babel/plugin-transform-dotall-regex'));
```
Comment (ide) [inline packages/babel-preset-expo/src/index.ts:116]: Doesn't Node LTS support this in V8?

Comment (EvanBacon) [reply packages/babel-preset-expo/src/index.ts]: Moving this change to another PR so I can evaluate it further.

**Review by ide — APPROVED** [review]

File: packages/@expo/cli/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- Enable package exports for server bundling. ([#24937](https://github.com/expo/expo/pull/24937) by [@EvanBacon](https://github.com/EvanBacon))
+
```
Comment (EvanBacon) [inline packages/@expo/cli/CHANGELOG.md:10]:
```suggestion
```

**Review by ide — APPROVED** [review]

**Pull request merged**